File: backend/src/services/vpnRotator.js
const axios = require('axios');
const logger = require('../utils/logger');

/**
 * VPN Rotation Service for YouTube API requests
 * Implements multi-provider VPN rotation to prevent IP blocking
 */
class VPNRotator {
  constructor() {
    this.enabled = process.env.VPN_ROTATION_ENABLED === 'true';
    this.providers = (process.env.VPN_PROVIDERS || '').split(',').filter(p => p.trim());
    this.currentProvider = null;
    this.currentEndpoint = null;
    this.rotationInterval = parseInt(process.env.VPN_ROTATION_INTERVAL) || 300000; // 5 minutes
    this.lastRotation = 0;
    this.requestCount = 0;
    this.maxRequestsPerEndpoint = 100;
    this.failureCount = new Map();
    this.maxFailures = 3;
    
    this.providerConfigs = {
      nordvpn: {
        apiKey: process.env.NORDVPN_API_KEY,
        endpoints: [
          'https://api.nordvpn.com/v1/servers/recommendations',
          'https://api.nordvpn.com/v1/servers'
        ],
        rotateMethod: 'api'
      },
      expressvpn: {
        apiKey: process.env.EXPRESSVPN_API_KEY,
        endpoints: [
          'https://www.expressvpn.com/api/v1/server/list'
        ],
        rotateMethod: 'api'
      },
      surfshark: {
        apiKey: process.env.SURFSHARK_API_KEY,
        endpoints: [
          'https://api.surfshark.com/v1/server/clusters'
        ],
        rotateMethod: 'api'
      },
      webshare: {
        username: process.env.WEBSHARE_USERNAME || 'enxguasp',
        password: process.env.WEBSHARE_PASSWORD || 'uthv5htk0biy',
        endpoints: [
          'rotating-residential.webshare.io:9000'
        ],
        rotateMethod: 'proxy'
      }
    };
    
    this.currentProxyList = [];
    this.currentProxyIndex = 0;
    
    if (this.enabled) {
      logger.info('VPN Rotator initialized', {
        enabled: this.enabled,
        providers: this.providers,
        rotationInterval: this.rotationInterval
      });
    }
  }

  /**
   * Check if VPN rotation should occur
   */
  shouldRotate() {
    if (!this.enabled) return false;
    
    const timeSinceLastRotation = Date.now() - this.lastRotation;
    const timeBasedRotation = timeSinceLastRotation >= this.rotationInterval;
    const requestBasedRotation = this.requestCount >= this.maxRequestsPerEndpoint;
    
    return timeBasedRotation || requestBasedRotation;
  }

  /**
   * Rotate to next VPN endpoint
   */
  async rotateVPN() {
    if (!this.enabled) return null;

    try {
      logger.info('Starting VPN rotation', {
        currentProvider: this.currentProvider,
        requestCount: this.requestCount,
        timeSinceLastRotation: Date.now() - this.lastRotation
      });

      // Select next provider
      const availableProviders = this.providers.filter(provider => 
        (this.failureCount.get(provider) || 0) < this.maxFailures
      );

      if (availableProviders.length === 0) {
        logger.error('All VPN providers have exceeded failure threshold');
        this.resetFailureCounts();
        return null;
      }

      // Round-robin provider selection (handle initial null state)
      let nextIndex = 0;
      if (this.currentProvider && availableProviders.includes(this.currentProvider)) {
        const currentIndex = availableProviders.indexOf(this.currentProvider);
        nextIndex = (currentIndex + 1) % availableProviders.length;
      }
      this.currentProvider = availableProviders[nextIndex];

      const config = this.providerConfigs[this.currentProvider];
      if (!config) {
        throw new Error(`No configuration found for provider: ${this.currentProvider}`);
      }

      // Rotate based on provider type
      let rotationResult;
      if (config.rotateMethod === 'proxy') {
        rotationResult = await this.rotateProxy(config);
      } else {
        rotationResult = await this.rotateVPNEndpoint(config);
      }

      this.lastRotation = Date.now();
      this.requestCount = 0;

      logger.info('VPN rotation completed', {
        provider: this.currentProvider,
        endpoint: this.currentEndpoint,
        success: !!rotationResult
      });

      return rotationResult;

    } catch (error) {
      logger.error('VPN rotation failed', {
        provider: this.currentProvider,
        error: error.message
      });
      
      this.recordFailure(this.currentProvider);
      throw error;
    }
  }

  /**
   * Rotate proxy endpoint (for proxy-based providers like Webshare)
   */
  async rotateProxy(config) {
    try {
      // For WebShare rotating residential proxy, use direct endpoint configuration
      if (this.currentProvider === 'webshare') {
        this.currentEndpoint = {
          type: 'proxy',
          host: config.endpoints[0].split(':')[0], // rotating-residential.webshare.io
          port: parseInt(config.endpoints[0].split(':')[1]), // 9000
          username: config.username,
          password: config.password,
          protocol: 'http'
        };

        logger.info('WebShare proxy endpoint configured', {
          host: this.currentEndpoint.host,
          port: this.currentEndpoint.port,
          username: this.currentEndpoint.username
        });

        return this.currentEndpoint;
      }

      // Legacy API-based proxy rotation for other providers
      if (this.currentProxyList.length === 0) {
        // Fetch fresh proxy list
        await this.fetchProxyList(config);
      }

      if (this.currentProxyList.length === 0) {
        throw new Error('No proxies available');
      }

      // Select next proxy
      this.currentProxyIndex = (this.currentProxyIndex + 1) % this.currentProxyList.length;
      const proxy = this.currentProxyList[this.currentProxyIndex];

      this.currentEndpoint = {
        type: 'proxy',
        host: proxy.proxy_address,
        port: proxy.port,
        username: proxy.username,
        password: proxy.password,
        protocol: proxy.protocol || 'http'
      };

      return this.currentEndpoint;

    } catch (error) {
      logger.error('Proxy rotation failed', { error: error.message });
      throw error;
    }
  }

  /**
   * Fetch proxy list from API (for legacy providers)
   * Note: WebShare now uses direct endpoint configuration
   */
  async fetchProxyList(config) {
    try {
      // Skip API call for WebShare - using direct endpoint
      if (this.currentProvider === 'webshare') {
        logger.info('WebShare uses direct endpoint configuration, skipping API call');
        return;
      }

      const response = await axios.get(config.endpoints[0], {
        headers: {
          'Authorization': `Token ${config.apiKey}`
        },
        timeout: 10000
      });

      if (response.data && response.data.results) {
        this.currentProxyList = response.data.results;
        this.currentProxyIndex = 0;
        
        logger.info('Fetched proxy list', { 
          provider: this.currentProvider,
          count: this.currentProxyList.length 
        });
      }

    } catch (error) {
      logger.error('Failed to fetch proxy list', { 
        provider: this.currentProvider,
        error: error.message 
      });
      throw error;
    }
  }

  /**
   * Rotate VPN endpoint (for traditional VPN providers)
   */
  async rotateVPNEndpoint(config) {
    try {
      // For traditional VPN providers, we would connect to their API
      // to get server recommendations and connect to a new server
      
      // This is a simplified implementation
      // In production, you would integrate with the actual VPN provider APIs
      
      const endpoint = config.endpoints[Math.floor(Math.random() * config.endpoints.length)];
      
      this.currentEndpoint = {
        type: 'vpn',
        provider: this.currentProvider,
        endpoint: endpoint
      };

      return this.currentEndpoint;

    } catch (error) {
      logger.error('VPN endpoint rotation failed', { error: error.message });
      throw error;
    }
  }

  /**
   * Get axios configuration with current VPN/proxy settings
   */
  getAxiosConfig() {
    if (!this.enabled || !this.currentEndpoint) {
      return {};
    }

    const config = {
      timeout: 30000
    };

    if (this.currentEndpoint.type === 'proxy') {
      config.proxy = {
        protocol: this.currentEndpoint.protocol,
        host: this.currentEndpoint.host,
        port: this.currentEndpoint.port,
        auth: {
          username: this.currentEndpoint.username,
          password: this.currentEndpoint.password
        }
      };

      // Add proxy headers
      config.headers = {
        'User-Agent': this.getRandomUserAgent()
      };
    }

    return config;
  }

  /**
   * Make HTTP request with VPN rotation
   */
  async makeRequest(url, options = {}) {
    if (!this.enabled) {
      return axios(url, options);
    }

    // Check if rotation is needed
    if (this.shouldRotate()) {
      await this.rotateVPN();
    }

    // Apply VPN configuration
    const vpnConfig = this.getAxiosConfig();
    const mergedConfig = { ...options, ...vpnConfig };

    try {
      this.requestCount++;
      const response = await axios(url, mergedConfig);
      
      // Reset failure count on successful request
      this.failureCount.set(this.currentProvider, 0);
      
      return response;

    } catch (error) {
      // Check if error is IP-related and should trigger rotation
      if (this.isIPBlockingError(error)) {
        logger.warn('IP blocking detected, forcing rotation', {
          provider: this.currentProvider,
          error: error.message
        });
        
        await this.rotateVPN();
        
        // Retry with new IP
        const newVpnConfig = this.getAxiosConfig();
        const retryConfig = { ...options, ...newVpnConfig };
        return axios(url, retryConfig);
      }
      
      throw error;
    }
  }

  /**
   * Check if error indicates IP blocking
   */
  isIPBlockingError(error) {
    if (!error.response) return false;
    
    const status = error.response.status;
    const blockingStatuses = [429, 403, 503];
    
    if (blockingStatuses.includes(status)) {
      return true;
    }
    
    const errorMessage = error.message.toLowerCase();
    const blockingKeywords = [
      'rate limit',
      'too many requests',
      'ip blocked',
      'access denied',
      'forbidden'
    ];
    
    return blockingKeywords.some(keyword => errorMessage.includes(keyword));
  }

  /**
   * Record failure for a provider
   */
  recordFailure(provider) {
    const currentFailures = this.failureCount.get(provider) || 0;
    this.failureCount.set(provider, currentFailures + 1);
    
    logger.warn('VPN provider failure recorded', {
      provider,
      failures: currentFailures + 1,
      maxFailures: this.maxFailures
    });
  }

  /**
   * Reset failure counts for all providers
   */
  resetFailureCounts() {
    this.failureCount.clear();
    logger.info('VPN provider failure counts reset');
  }

  /**
   * Get random user agent to avoid detection
   */
  getRandomUserAgent() {
    const userAgents = [
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
      'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0',
      'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:89.0) Gecko/20100101 Firefox/89.0'
    ];
    
    return userAgents[Math.floor(Math.random() * userAgents.length)];
  }

  /**
   * Get current status
   */
  getStatus() {
    return {
      enabled: this.enabled,
      currentProvider: this.currentProvider,
      currentEndpoint: this.currentEndpoint,
      requestCount: this.requestCount,
      lastRotation: this.lastRotation,
      timeSinceLastRotation: Date.now() - this.lastRotation,
      availableProviders: this.providers,
      failureCounts: Object.fromEntries(this.failureCount)
    };
  }

  /**
   * Force rotation (for testing/manual triggers)
   */
  async forceRotation() {
    this.lastRotation = 0; // Force time-based rotation
    this.requestCount = this.maxRequestsPerEndpoint; // Force request-based rotation
    return this.rotateVPN();
  }

  /**
   * Disable VPN rotation
   */
  disable() {
    this.enabled = false;
    logger.info('VPN rotation disabled');
  }

  /**
   * Enable VPN rotation
   */
  enable() {
    this.enabled = true;
    logger.info('VPN rotation enabled');
  }
}

module.exports = VPNRotator;

File: ml-services/transcript-processor/src/vpn_rotator.py
"""
VPN Rotation Service for Python transcript extraction
Handles IP rotation using WebShare proxy service
"""
import asyncio
import aiohttp
import time
import random
from typing import Dict, List, Optional, Any
from dataclasses import dataclass
import structlog

from .config import settings

logger = structlog.get_logger(__name__)

@dataclass
class ProxyEndpoint:
    """Proxy endpoint configuration"""
    host: str
    port: int
    username: str
    password: str
    protocol: str = "http"
    last_used: float = 0
    failure_count: int = 0

class VPNRotator:
    """VPN/Proxy rotation service using WebShare direct credentials"""
    
    def __init__(self):
        self.enabled = settings.enable_vpn_rotation
        self.webshare_username = settings.webshare_proxy_username
        self.webshare_password = settings.webshare_proxy_password
        self.rotation_interval = settings.proxy_rotation_interval
        self.max_failures = 3
        self.current_proxy_index = 0
        self.proxy_list: List[ProxyEndpoint] = []
        self.last_rotation = 0
        self.request_count = 0
        self.max_requests_per_proxy = 50
        self.session = None
        
        # Create the proxy endpoint directly from credentials
        if self.enabled and self.webshare_username and self.webshare_password:
            # WebShare rotating residential proxy endpoint
            self.proxy_list = [
                ProxyEndpoint(
                    host="rotating-residential.webshare.io",
                    port=9000,
                    username=self.webshare_username,
                    password=self.webshare_password,
                    protocol="http"
                )
            ]
            logger.info("VPN Rotator initialized with WebShare direct credentials",
                       rotation_interval=self.rotation_interval,
                       max_failures=self.max_failures,
                       proxy_endpoint="rotating-residential.webshare.io:9000")
        elif self.enabled:
            logger.error("VPN rotation enabled but WebShare credentials not provided")
            self.enabled = False
    
    async def initialize(self):
        """Initialize the VPN rotator"""
        if not self.enabled:
            return
        
        # Direct credentials mode - already initialized in __init__
        if self.proxy_list:
            logger.info("VPN Rotator initialization complete", 
                       proxy_count=len(self.proxy_list))
            return
        
        # Fallback to API mode if no direct credentials
        try:
            await self.fetch_proxy_list()
            logger.info("VPN Rotator initialization complete", 
                       proxy_count=len(self.proxy_list))
        except Exception as e:
            logger.error("Failed to initialize VPN rotator", error=str(e))
            self.enabled = False
    
    async def fetch_proxy_list(self):
        """Fetch proxy list from WebShare API (legacy method)"""
        # Skip if using direct credentials
        if self.proxy_list:
            logger.info("Using direct WebShare credentials, skipping API call")
            return
            
        raise Exception("API key method not supported with direct credentials")
        
        try:
            headers = {
                'Authorization': f'Token {self.api_key}',
                'Content-Type': 'application/json'
            }
            
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    'https://proxy.webshare.io/api/v2/proxy/list/',
                    headers=headers,
                    timeout=aiohttp.ClientTimeout(total=30)
                ) as response:
                    if response.status != 200:
                        raise Exception(f"WebShare API error: {response.status}")
                    
                    data = await response.json()
                    
                    if 'results' not in data:
                        raise Exception("Invalid WebShare API response")
                    
                    self.proxy_list = []
                    for proxy_data in data['results']:
                        proxy = ProxyEndpoint(
                            host=proxy_data['proxy_address'],
                            port=proxy_data['port'],
                            username=proxy_data['username'],
                            password=proxy_data['password'],
                            protocol='http'
                        )
                        self.proxy_list.append(proxy)
                    
                    # Shuffle the list for random distribution
                    random.shuffle(self.proxy_list)
                    
                    logger.info("Fetched proxy list from WebShare", 
                               count=len(self.proxy_list))
        
        except Exception as e:
            logger.error("Failed to fetch proxy list", error=str(e))
            raise
    
    def should_rotate(self) -> bool:
        """Check if proxy should be rotated"""
        if not self.enabled or not self.proxy_list:
            return False
        
        # Time-based rotation
        time_since_rotation = time.time() - self.last_rotation
        if time_since_rotation >= self.rotation_interval:
            return True
        
        # Request-based rotation
        if self.request_count >= self.max_requests_per_proxy:
            return True
        
        # Failure-based rotation
        if self.proxy_list:
            current_proxy = self.proxy_list[self.current_proxy_index]
            if current_proxy.failure_count >= self.max_failures:
                return True
        
        return False
    
    async def rotate_if_needed(self):
        """Rotate proxy if needed"""
        if not self.should_rotate():
            return
        
        await self.rotate_proxy()
    
    async def rotate_proxy(self):
        """Rotate to next available proxy"""
        if not self.enabled or not self.proxy_list:
            return
        
        original_index = self.current_proxy_index
        attempts = 0
        max_attempts = len(self.proxy_list)
        
        while attempts < max_attempts:
            # Move to next proxy
            self.current_proxy_index = (self.current_proxy_index + 1) % len(self.proxy_list)
            current_proxy = self.proxy_list[self.current_proxy_index]
            
            # Skip proxies with too many failures
            if current_proxy.failure_count < self.max_failures:
                current_proxy.last_used = time.time()
                self.last_rotation = time.time()
                self.request_count = 0
                
                logger.info("Rotated to new proxy", 
                           index=self.current_proxy_index,
                           host=current_proxy.host,
                           port=current_proxy.port)
                return
            
            attempts += 1
        
        # All proxies have failures - refresh the list
        logger.warn("All proxies have failures, refreshing proxy list")
        try:
            await self.fetch_proxy_list()
            self.current_proxy_index = 0
            if self.proxy_list:
                self.last_rotation = time.time()
                self.request_count = 0
        except Exception as e:
            logger.error("Failed to refresh proxy list", error=str(e))
    
    def get_current_proxy(self) -> Optional[ProxyEndpoint]:
        """Get current proxy configuration"""
        if not self.enabled or not self.proxy_list:
            return None
        
        return self.proxy_list[self.current_proxy_index]
    
    def get_proxy_config(self) -> Optional[Dict[str, Any]]:
        """Get proxy configuration for HTTP requests"""
        proxy = self.get_current_proxy()
        if not proxy:
            return None
        
        return {
            'proxy': f'{proxy.protocol}://{proxy.username}:{proxy.password}@{proxy.host}:{proxy.port}',
            'proxy_auth': aiohttp.BasicAuth(proxy.username, proxy.password)
        }
    
    async def make_request(self, method: str, url: str, **kwargs) -> aiohttp.ClientResponse:
        """Make HTTP request with proxy rotation"""
        if not self.enabled:
            # Make request without proxy
            async with aiohttp.ClientSession() as session:
                return await session.request(method, url, **kwargs)
        
        # Check if rotation is needed
        await self.rotate_if_needed()
        
        proxy_config = self.get_proxy_config()
        if proxy_config:
            kwargs.update(proxy_config)
        
        # Add random user agent
        if 'headers' not in kwargs:
            kwargs['headers'] = {}
        kwargs['headers']['User-Agent'] = self.get_random_user_agent()
        
        try:
            self.request_count += 1
            
            async with aiohttp.ClientSession() as session:
                response = await session.request(method, url, **kwargs)
                
                # Reset failure count on successful request
                if self.proxy_list and response.status < 400:
                    current_proxy = self.proxy_list[self.current_proxy_index]
                    current_proxy.failure_count = 0
                
                return response
        
        except Exception as e:
            # Record failure and potentially rotate
            if self.proxy_list:
                current_proxy = self.proxy_list[self.current_proxy_index]
                current_proxy.failure_count += 1
                
                logger.warning("Request failed with current proxy", 
                              error=str(e),
                              proxy_host=current_proxy.host,
                              failure_count=current_proxy.failure_count)
                
                # Try rotating and retrying once
                if current_proxy.failure_count >= self.max_failures:
                    await self.rotate_proxy()
                    
                    # Retry with new proxy
                    proxy_config = self.get_proxy_config()
                    if proxy_config:
                        retry_kwargs = kwargs.copy()
                        retry_kwargs.update(proxy_config)
                        
                        async with aiohttp.ClientSession() as session:
                            return await session.request(method, url, **retry_kwargs)
            
            raise
    
    def get_random_user_agent(self) -> str:
        """Get random user agent string"""
        user_agents = [
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:121.0) Gecko/20100101 Firefox/121.0',
            'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:121.0) Gecko/20100101 Firefox/121.0'
        ]
        return random.choice(user_agents)
    
    def record_success(self):
        """Record successful request"""
        if self.proxy_list:
            current_proxy = self.proxy_list[self.current_proxy_index]
            current_proxy.failure_count = max(0, current_proxy.failure_count - 1)
    
    def record_failure(self):
        """Record failed request"""
        if self.proxy_list:
            current_proxy = self.proxy_list[self.current_proxy_index]
            current_proxy.failure_count += 1
    
    def get_status(self) -> Dict[str, Any]:
        """Get VPN rotator status"""
        current_proxy = self.get_current_proxy()
        
        return {
            'enabled': self.enabled,
            'proxy_count': len(self.proxy_list),
            'current_proxy_index': self.current_proxy_index,
            'current_proxy': {
                'host': current_proxy.host if current_proxy else None,
                'port': current_proxy.port if current_proxy else None,
                'failure_count': current_proxy.failure_count if current_proxy else 0
            } if current_proxy else None,
            'request_count': self.request_count,
            'last_rotation': self.last_rotation,
            'time_since_rotation': time.time() - self.last_rotation,
            'rotation_needed': self.should_rotate()
        }
    
    async def force_rotation(self):
        """Force proxy rotation (for testing)"""
        await self.rotate_proxy()
    
    async def refresh_proxies(self):
        """Refresh proxy list from WebShare"""
        await self.fetch_proxy_list()
    
    async def cleanup(self):
        """Cleanup resources"""
        if self.session and not self.session.closed:
            await self.session.close()